File: src/pulseq_reports/assets/cards/diagram.js
// Sequence diagram card: one lane chart with a button for each time window that
// cards/diagram.py gives. Unlike the old card, there are no lane sets and no point
// budget: cards/diagram.py sends the compressed block and event tables of each file
// that at least one window uses (diagram_data.diagram_tables), and this script
// decodes them (base64 -> gzip -> typed arrays -> SeqLanes.decode) into a model for
// each file, once, on first use. The chart's `lanesFor` hook (lane_chart.js) then
// asks SeqLanes.lanesFor for the lanes of the current view on every render: the exact
// waveform when the view has few enough points, or the minimum and the maximum of
// each lane in each of the plot's time bins otherwise. The status line under the
// chart (`{card_id}-mode`) says which of the two the current render shows.
//
// A file's tables can be large (up to 10^7 blocks), so decoding runs only for the
// file of the first window at start, and for another file the first time a button
// selects one of its windows; each decoded model is kept in `models` (by file
// index), so a file already shown is not decoded twice. While a file decodes, the
// status line shows "Loading..." and the window buttons are disabled.
PulseqReport.registerCard("diagram", async (section, data) => {
  const buttons = section.querySelectorAll("[data-window]");
  const statusEl = document.getElementById(`${section.id}-mode`);
  const files = data.files;
  const windows = data.windows;
  const models = new Array(files.length).fill(null);

  function setButtonsDisabled(disabled) {
    for (const button of buttons) button.disabled = disabled;
  }

  function showStatus(exact, bins) {
    statusEl.textContent = exact
      ? "Exact waveform."
      : `Minimum and maximum in each of ${bins} time bins. Zoom in to see the exact waveform.`;
  }

  // The typed array for one table entry {dtype, length, data}: base64 -> gzip bytes
  // -> DecompressionStream -> the typed array for its dtype. Throws for a dtype this
  // script does not know, and when the decompressed length does not match `length`
  // (a sign that the table or the decode is wrong).
  async function decodeTable(entry) {
    // A plain loop: the text can be tens of MB, and a callback for each character
    // (Uint8Array.from with a map function) is much slower.
    const text = atob(entry.data);
    const bytes = new Uint8Array(text.length);
    for (let i = 0; i < text.length; i++) bytes[i] = text.charCodeAt(i);
    const buffer = await new Response(
      new Blob([bytes]).stream().pipeThrough(new DecompressionStream("gzip"))
    ).arrayBuffer();
    const ctor = {
      uint8: Uint8Array,
      uint16: Uint16Array,
      uint32: Uint32Array,
      float64: Float64Array,
    }[entry.dtype];
    if (!ctor) throw new Error(`diagram card: unknown table dtype "${entry.dtype}"`);
    const array = new ctor(buffer);
    if (array.length !== entry.length) {
      throw new Error(
        `diagram card: table length ${array.length} does not match the given length ` +
          `${entry.length}`
      );
    }
    return array;
  }

  // The decoded SeqLanes model of file `fileIndex`, from the cache when it is
  // already there. Decodes every one of the file's tables in parallel.
  async function decodeFile(fileIndex) {
    if (models[fileIndex]) return models[fileIndex];
    const file = files[fileIndex];
    const names = Object.keys(file.tables);
    const arrays = await Promise.all(names.map(name => decodeTable(file.tables[name])));
    const tables = {};
    names.forEach((name, i) => {
      tables[name] = arrays[i];
    });
    const model = SeqLanes.decode(data.format, tables, file.lanes);
    models[fileIndex] = model;
    return model;
  }

  // The window whose view is the chart's current xDomain (what a zoom-control Reset
  // returns to). Only a button that switches to another file's model moves the
  // xDomain, so only that kind of click updates `initialWindow`; a same-file click
  // (setView) leaves it as is.
  let initialWindow = 0;
  let currentFileIndex = windows[0].file;

  statusEl.textContent = "Loading…";
  setButtonsDisabled(true);
  // The first file's decode failure is not caught here: it propagates out of this
  // async init, so page.js shows the card's usual "could not be drawn" note (there
  // is no chart to fall back to without a first model).
  let current = await decodeFile(currentFileIndex);
  setButtonsDisabled(false);

  function press(i) {
    for (const button of buttons) {
      button.setAttribute("aria-pressed", String(Number(button.dataset.window) === i));
    }
  }

  const chart = PulseqReport.laneChart({
    svg: document.getElementById(`${section.id}-diagram`),
    chart: document.getElementById(`${section.id}-chart`),
    tip: document.getElementById(`${section.id}-tip`),
    // Only its length (the number of lanes) matters here: lane_chart.js reads it for
    // the SVG height before the first render, which then replaces it with the
    // provider's own lanes. `current.lanesMeta` already has the right count, so
    // building it needs no SeqLanes.lanesFor call.
    lanes: current.lanesMeta,
    lanesFor: (view, bins) => {
      const r = SeqLanes.lanesFor(current, view, bins);
      showStatus(r.exact, bins);
      return r.lanes;
    },
    xDomain: windows[0].view_ms,
    extent: [0, current.durationS * 1000],
    minSpan: 0.01,
    xLabel: "Time (ms)",
    cursorText: v => `t = ${v.toFixed(3)} ms`,
    // A zoom or pan leaves no button pressed; a reset (isInitial) presses the button
    // of the window whose view is the current xDomain.
    onViewChange: (view, isInitial) => {
      if (isInitial) {
        press(initialWindow);
      } else {
        for (const button of buttons) button.setAttribute("aria-pressed", "false");
      }
    },
  });

  for (const button of buttons) {
    const i = Number(button.dataset.window);
    button.addEventListener("click", async () => {
      const w = windows[i];
      if (w.file === currentFileIndex) {
        press(i);
        chart.setView(w.view_ms);
        return;
      }
      setButtonsDisabled(true);
      statusEl.textContent = "Loading…";
      let model;
      try {
        model = await decodeFile(w.file);
      } catch (error) {
        // A later file's decode failure is shown in the status line instead of
        // being rethrown: the chart already has a model and a view to keep showing,
        // so the card stays usable instead of being replaced by the "could not be
        // drawn" note.
        statusEl.textContent = `Could not load "${files[w.file].name}": ${error.message}`;
        setButtonsDisabled(false);
        return;
      }
      press(i);
      current = model;
      currentFileIndex = w.file;
      // As for the first render, only the number of lanes is read before the render
      // replaces them with the provider's lanes.
      chart.setWindow({
        lanes: current.lanesMeta,
        xDomain: w.view_ms,
        extent: [0, current.durationS * 1000],
      });
      initialWindow = i;
      setButtonsDisabled(false);
    });
  }
});

File: src/pulseq_reports/cards/diagram.py
"""Sequence diagram card: RF, ADC and gradient waveforms against time, with one button
for each time window that the caller gives.

The card sends the compressed block and event tables of each file that at least one
window uses (`diagram_data.diagram_tables`, `encode_tables`), not expanded points. The
browser (`assets/cards/diagram.js`, `assets/seq_lanes.js`) decodes them and draws the
exact waveform when a view has few enough points, or the minimum and the maximum of
each lane in each of the plot's time bins otherwise, so the card works for a file of up
to 10^7 blocks (`docs/plans/diagram-event-table.md`).
"""

import html
from collections.abc import Sequence

from ..diagram_data import diagram_tables, encode_tables, lane_meta
from ..extensions import refuse_rotations
from ..markup import _zoom_controls
from ..page import Card
from ..seq_utils import NamedSequence
from ..waveforms import TimeWindow, duration_s


def _ms(t_s: float) -> float:
    return round(t_s * 1e3, 4)


def _validate(seqs: Sequence[NamedSequence], windows: Sequence[TimeWindow]) -> None:
    if not windows:
        raise ValueError("the diagram card needs at least one time window")
    for w in windows:
        if not 0 <= w.file_index < len(seqs):
            raise ValueError(f"window {w.label!r}: file_index {w.file_index} is not a file")
        if not w.end_s > w.start_s:
            raise ValueError(f"window {w.label!r}: the end is not after the start")


def _diagram_data(seqs: Sequence[NamedSequence], windows: Sequence[TimeWindow]) -> dict:
    """The card data (section 4.1 of `docs/plans/diagram-event-table.md`):
    `{"format": 1, "files": [...], "windows": [...]}`.

    `files` has one entry for each file that at least one window uses, in the order
    in which `windows` first use them (so the file of the first window is entry 0):
    `name`, `duration_s`, `num_blocks`, `lanes` (`lane_meta`) and `tables`
    (`encode_tables(diagram_tables(seq))`). `diagram_tables(seq)` is built once for
    each such file, and passed to `lane_meta` so it is not built twice. `windows` has
    one entry for each of `windows`: `label`, `file` (the index into `files`, not into
    `seqs`) and `view_ms`.
    """
    file_index: dict[int, int] = {}  # index into seqs -> index into "files"
    files: list[dict] = []
    out_windows: list[dict] = []

    for w in windows:
        if w.file_index not in file_index:
            seq = seqs[w.file_index].seq
            tables = diagram_tables(seq)
            file_index[w.file_index] = len(files)
            files.append(
                {
                    "name": seqs[w.file_index].name,
                    "duration_s": duration_s(seq),
                    "num_blocks": len(seq.block_events),
                    "lanes": lane_meta(seq, tables),
                    "tables": encode_tables(tables),
                }
            )
        named = seqs[w.file_index]
        label = w.label if len(seqs) == 1 else f"{named.name}: {w.label}"
        out_windows.append(
            {
                "label": label,
                "file": file_index[w.file_index],
                "view_ms": [_ms(w.start_s), _ms(w.end_s)],
            }
        )
    return {"format": 1, "files": files, "windows": out_windows}


def diagram_card(
    seqs: Sequence[NamedSequence],
    windows: Sequence[TimeWindow],
    card_id: str = "diagram",
) -> Card:
    """The "Sequence diagram" card: one button for each of `windows`, in the given
    order. The first window is the initial view. With more than one file, each button
    text starts with the file name.

    Each file that at least one window uses gets its block and event tables sent to
    the browser once (see `_diagram_data`); the browser decodes them and draws the
    waveform. A status line under the chart (`{card_id}-mode`) says whether the
    current view is exact or a minimum/maximum of time bins.

    Raises `ValueError` when `windows` is empty, or a window names no file or ends
    before it starts. Raises `NotImplementedError` (`extensions.refuse_rotations`) for
    a sequence that uses the Pulseq rotation extension: this card shows unrotated
    gradient events, which would be wrong for such a sequence.
    """
    _validate(seqs, windows)
    for named in seqs:
        refuse_rotations(named.seq)
    data = _diagram_data(seqs, windows)
    buttons = "".join(
        f'<button type="button" data-window="{i}" aria-pressed="{str(i == 0).lower()}">'
        f"{html.escape(w['label'])}</button>"
        for i, w in enumerate(data["windows"])
    )
    svg_id = f"{card_id}-diagram"
    body = (
        f'<div class="controls" role="group" aria-label="Time window">{buttons}</div>\n'
        + _zoom_controls(svg_id)
        + f'\n<div class="chart" id="{card_id}-chart">\n'
        f'<svg id="{svg_id}" tabindex="0" role="img"\n'
        '  aria-label="Sequence diagram: RF magnitude and phase, ADC, Gx, Gy and Gz against '
        'time"></svg>\n'
        f'<div class="tip" id="{card_id}-tip" hidden></div>\n'
        "</div>\n"
        f'<p class="muted" id="{card_id}-mode" aria-live="polite">Loading…</p>\n'
        '<p class="muted">Hover the diagram, or focus it and use the arrow keys, to read '
        "values.\nRF phase is drawn where |B1| is above 1% of that pulse's peak.\n"
        "Click the chart to mark the centre for the zoom buttons. Drag across the chart to "
        "zoom to that range. Hold Shift and drag, or scroll sideways, to pan.</p>"
    )
    return Card(id=card_id, title="Sequence diagram", body_html=body, data=data, script="diagram")
